### ISSUE TITLE
Feature request: Make it possible to change a username

### ISSUE THREAD

**Issue by jeroenpraat** [tootsuite/mastodon#15320]
I know it's technically difficult, but still it's very important for trans and non-binary people to be able to change their username. For especially trans people it forces them to use their deadname or they are forced to make a new account. In either way they are forced. 

Maybe if implemented, it should maybe be limited to e.g. 1 per 3 months. On the birdsite it's possible and it is one of the reasons people stay there.

**Comment by mashirozx** [comment]
Why not creat a new account and then alias the old to the new?

**Comment by jeroenpraat** [comment]
@mashirozx Because people don't want to refer to their old name / deadname.

**Comment by mashirozx** [comment]
No, it's not refer to old name, but refer to the new name in old account.

**Comment by jeroenpraat** [comment]
@mashirozx People want to be able to refer to old content. So they still need to mention the old account. And still they are forced to make this somehow work (create new account, refer to both), only because they are trans or non-binary. That's transphobia and nb-phobia. They just want to keep their account, with all content, but with a different name.

**Comment by ThibG** [comment]
Unfortunately, all identifiers, for accounts and statuses, include the username. So whichever solution we end up choosing, we'll have to deal with either of those issues:
- old links and references to existing posts are broken
- even though a handle change is allowed, the identifiers continue to use the old handle (e.g., `@bar@domain.com` would be served by `https://domain.com/users/foo`

Another possibility would be to stop using the username in **new** identifiers, but the issue would remain for **older** identifiers.

Additionally, whatever we do, current versions of Mastodon won't be able to understand those name changes, so it'll be a while until that feature is possible.

**Comment by jeroenpraat** [comment]
@ThibG Why not use the public key in the account table as a unique identifier? Or create one. I'm not a developer, I just try to make this huge problem visible.

**Comment by ThibG** [comment]
The issue isn't really about creating a new identifier, but migrating identifiers. Let me delve into this a bit deeper: in Mastodon, users basically have two “public” identifiers:
- the webfinger `acct:` URI: `foo@domain.com`
- the ActivityPub identifier: `https://domain.com/users/foo`

Currently, Mastodon uses the username in both identifiers. From a federation perspective, the ActivityPub identifier isn't supposed to change at all: if you change identifiers, you're another account altogether. Currently, Mastodon doesn't have any mechanism to handle one of those identifiers changing: trying to do so will either break stuff or keep using the old identifier.

So, whatever we chose to do, current Mastodon instances won't handle this properly when a user from a newer instance changes identifiers.

Furthermore, even if we do change identifiers, every old toot will be known by other instances with their old identifiers, which include the old username. Either we keep them forever (which isn't trivial), either we change their identifier and break federated interaction with them (that is, [un]boosts, [un]favs, replies, etc.)

It is likely that we will soon be able to support changing either account identifier without breaking everything, but even if we do so, it'll still be a long way before proper federated username change without breaking stuff, there are lots of open questions with that.

**Comment by jeroenpraat** [comment]
I appreciate your explanation @ThibG. But isn't it a fact that how longer we wait with adding those new extra identifiers, how longer it takes that username can be changed?

It's OK if it is a long process. As far this can be explained to people suffering from this limitation. The process could have been started already a year or two years ago, but I'm not here to blame people.

**Comment by trwnh** [comment]
>Another possibility would be to stop using the username in new identifiers, but the issue would remain for older identifiers.

#10745

**Comment by ThibG** [comment]
@jeroenpraat depends on how we end up doing it. A safe way to do it would be to use username-independent identifiers for **new** accounts and statuses. Of course, that would mean that older accounts would not be able to benefit from the feature.

Other alternatives require finding out a way to move from an ActivityPub identifier to another while breaking the least amount of stuff. The sooner we find a way to do that, the sooner the feature could be available, but switching to new identifiers for everything before that is just going to break stuff. And again, there is a logical hurdle here: the ActivityPub identifier is supposed to uniquely identify your account and never change. If you change identifiers, you change accounts.

**Comment by jeroenpraat** [comment]
@ThibG But why not using two identifiers for the time being?

**Comment by ThibG** [comment]
@jeroenpraat because there can only be one!

**Comment by trwnh** [comment]
> finding out a way to move from an ActivityPub identifier to another while breaking the least amount of stuff.
- HTTP 301 from old to new, 302 for username changes
- ActivityPub Move activity from old to new

pretty sure mastodon supports the latter, but does nothing with 301/302, right?

>The sooner we find a way to do that, the sooner the feature could be available, but switching to new identifiers for everything before that is just going to break stuff.

step 1 is to start assigning new activitypub "id" that doesn't depend on username, as pointed out in #10745

>the ActivityPub identifier is supposed to uniquely identify your account and never change. If you change identifiers, you change accounts.

nightpool pointed out to me once that there's nothing in the spec preventing changing "id", even if it does break stuff. i think as long as we accept that activitypub is currently bound to http (GET and POST are part of the spec, after all), then http 301/302 support shouldn't be too far out. #8465 asks for this as well.

**Comment by ThibG** [comment]
> > finding out a way to move from an ActivityPub identifier to another while breaking the least amount of stuff.
> 
>     * HTTP 301 from old to new, 302 for username changes

I don't think that's enough: consider the case where a toot moved from `https://example.com/users/foo/statuses/12345` to `https://example.com/statuses/12345`, with a remote instance A knowing about the toot with the former identifier, and instance B knowing it with the new identifier. Someone on A follows someone on B who ends up boosting the toot. A has no way of knowing that the newly-discovered toot at `https://example.com/statuses/12345` is indeed the same as `https://example.com/users/foo/statuses/12345` and will create a duplicate toot. Furthermore, if the original author wants to delete their toot, they'll send a `Delete` with the new id as a target, so A won't delete the toot.

In the other direction (instance learning about the old identifier while the new one is in use), a 301 or 302 might work, depending on how it's handled: I'm not sure what ActivityPub says exactly about referring to an object with an URL that it is not actually its identifier (but is enough to reach it). I'm not sure how the various implementations actually work with this either, although Mastodon might behave like you'd want it to do here.

>     * ActivityPub Move activity from old to new

The meaning of the `Move` activity to Mastodon is basically “this person has moved from account A to account B, but those two accounts exist and are distinct”. In particular, toots are not re-assigned to the account being moved to, and neither are follow relationships. Mastodon does re-issue follow requests, though, which is helpful, but not as seamless as a migration not explicitly requested by the user should be.

Either you're happy with this handling, and thus you're happy with the existing migration feature and don't really need that account renaming thing, or that `Move` handler isn't enough and we need to do *something* differently and risk having to special-case stuff in a way that isn't clear at all to me.

Furthermore, your suggested migration path seem to imply that on update, a Mastodon instance would send `Move` activities for each of its managed accounts to basically the whole fediverse.

> > The sooner we find a way to do that, the sooner the feature could be available, but switching to new identifiers for everything before that is just going to break stuff.
> 
> step 1 is to start assigning new activitypub "id" that doesn't depend on username, as pointed out in #10745

step 1 to what? step 1 to better handling changes to ActivityPub identifiers would be to immediately change identifiers without a way to have the change not break things?

As I said earlier, it seems reasonable to use identifiers that don't depend on username for **new** objects (at the cost of rewriting quite a bit of Mastodon, but that's not a fundamental issue), but not for existing ones.

**Comment by jeroenpraat** [comment]
> @jeroenpraat because there can only be one! Every object has exactly one identifier, that is supposed to be unique and not change. There's only place for one, and if you change it, any instance that knew about the old one is going to be confused (in most cases: think it's an entirely new object)

What I mean is to have two, one active and one non-active for future use.

**Comment by ThibG** [comment]
> What I mean is to have two, one active and one non-active for future use.

Yes but no, that doesn't make more sense, you can't declare an alternate “inactive” identifier for future use. Creating a new identifier isn't an issue (we could re-use the internal numerical identifier), the issue is that there isn't any mechanism for an object (toot, account) to change identifiers.

**Comment by Packbat** [comment]
I'm glad people are discussing the complexities of implementing this kind of feature, but I'd like to hop in and encourage people to approach this the way they'd approach, say, an issue that causes the server to irregularly crash: it's a pretty big deal and can make it impossible for people to participate, so the question is "what issues does this have, which ones can we mitigate, and which ones can we live with?"

The problem of zombie posts under the old account name is a real one - would it be better to send a specially-formatted "reassign this post's account" message that old instances interpret as a delete request, give federated instances a ledger of name changes that they apply to every post as it comes in (oh god that sounds like it could be really processor-intensive), resend all the old posts with the new account (oof and also is that even possible and would it spam timelines?), let them lurch around all deadname-y and gross ... this is a good question, and I think we should approach it with the understanding that we should be doing *something*.

Because people's names do change, and it's reasonable to expect people to be able to change what they are called, and - as the OP points out - it's an accessibility issue for a large population of users. So I don't think it should be dismissed as too-hard - I think it should be approached as painful but necessary.

**Comment by jeroenpraat** [comment]
I'm not going into any technical details, because I'm simply not a programmer. I know my place. :wink: The only thing I want to add is:

**_Where there is a will, there's a way._**

**Comment by Packbat** [comment]
...actually, would it make sense to have a legacy setting on the change-username page to choose between "migrate/delete old posts" and "migrate/leave unchanged old posts"? It would allow people changing usernames for less weighty reasons to preserve as much of their old content as possible in the wild, while allowing the trans folks referenced in the OP to remove as much deadnaming content as feasible.

**Comment by phseiff** [comment]
The way I understand it, the closest we have to this right now is migrating to a new account on a different instance, and linking the old account to the new one.
Why not build upon that?

I think there could be two ways of account redirects, one where the redirecting account is visible to everyone (the thing we have right now with people switching to another instance), and another one where it is invisible, but still existing in the backend, and every mention of the old account is internally interpreted as/ converted to a mention of the new account (so mentions of the old account are internally re-directed to the inbox of the new account and the URL of the old account to the new account, without actually modifying any of the toots that contains such mentions).

This would be compatible with old instances, since it would all happen on the side of the instance in which the person who changed their username resides, and no posts mentioning the person would need to be modified.

The whole thing could also involve transferring the posts and DMs of the old account to the new account, as well as its password and things like that, which I assume would not be too difficult since it is pretty much just a database thing (note that this is easier than migrating an account to another instance, since it's not a cross-instance thing).

Since the whole thing would leave an orphaned unusable username behind, it's probably best to tie it to a verification process by the admin (as in, the Admin needs to approve of it in the Admin dashboard), but this shouldn't be too much of a problem assuming the admins are reasonable and not transphobic/enbyphobic.

**Comment by CodingCaro** [comment]
M

**Comment by meg-gutshall** [comment]
I know this is a huge, technically complicated issue, but it's important **and** I just learned that Mastodon is built on Ruby!! 😍 

I'm commenting to follow this issue and help code a workable solution. I'd be happy to pitch in on this issue along with someone who may be more knowledgeable of the codebase! I haven't made any PRs to mastodon before.

**Comment by vieko** [comment]
I know it is silly, but I just want to change the capitalization of my handle from @Vieko to @vieko, that's all!

**Comment by gjvnq** [comment]
I have one idea: what if we kinda copy git's mailmap system?

From the "core system's" perspective there are two users `@deadname` and `@newname` with the former being a redirect to the later.

From the UI perspective, the `@deadname` is an account with a special "tag" saying "please, replace me with `@newname` when showing to the user". This has the advantage of keeping compatibility but the disadvantage of showing the deadname handle on incompatible systems.

I'm not very knowledgeable on the internals of ActivityPub so I don't know how hard it is to implement that tagging thing without causing compatibility issues.

I will try¹ to take a look at the ActivityPub spec to see how best to handle this. As for implementation, I'm a somewhat skilled programmer but I have very little experience with Ruby so I don't know if I will be able to implement this.

¹ It's nearing the end of the semester for me so I tend to get busier than usual.

**Comment by rokejulianlockhart** [comment]
https://git.pleroma.social/pleroma/pleroma/-/issues/59#note_97208
https://github.com/pixelfed/pixelfed/issues/4053

**Comment by rokejulianlockhart** [comment]
Would merely automating the redirection and transferral process – what https://github.com/pixelfed/pixelfed/issues/4053#issuecomment-1407708709 describes – be sufficient, or actually, superior?

**Comment by j-f1** [comment]
Would it be possible to change the ActivityPub ID for new users to the numerical ID, thus potentially enabling them to change usernames in the future even if existing users will require more work to be done to ensure that they can change their username?

**Comment by trwnh** [comment]
@j-f1 yes, exactly... #10745

**Comment by AmkiTakk** [comment]
Commenting to keep an eye on how this evolves. In a past instance, I made a typo writing my username "AmkITakk" instead of "AmkiTakk" which was pretty embarrassing. It's fixed in my new instance but I'd like to change my username to something completely different now and I can't without making a new account (considering all the art I'd have to repost, it's just not doable).
Discord had a nice system (rip), where users had a unique identifier that was just numbers, and the actual username was freely modifiable. A similar system would be fantastic, and is highly requested all over the Fediverse as previous comments have shown!

**Comment by jcklpe** [comment]
commenting to follow the development of this feature (if it should ever go forward, which I hope it would (or at least something better than the current redirect solution, which is confusing for laypersons. Maybe nomadic or portable identity would be helpful in the fediverse?)

**Comment by coldReactive** [comment]
I'm commenting here to say that I'm not quite on board with allowing people to change their usernames "willy-nilly." I've had people I've seen on twitter change their username every few months, most of these same people did it on Discord too before Discord changed how they handled usernames. It often comes down to either a petty reason, or because their account is actually an RP account and they wish to change characters or similar.

For the latter, I would think it would be better suited for a specific host/instance of mastodon to be able to offer that kind of service, rather than force it upon all of the instances.

**Comment by tedajax** [comment]
This needs to be a feature today and the fact that it's not leads me to believe mastodon is not a serious project

**Comment by ClearlyClaire** [comment]
This is on our radar, but there are at least two technical hurdles:
- the protocol identifier of every actor and post currently includes the username, and can't be retroactively changed, this means we won't be able to solve the issue, or at least not in a clean and efficient way, for existing accounts; we plan to move to another protocol identifier scheme for new accounts, but making sure the two schemes coexist without breaking anything will be quite some work
- Mastodon uses the `username@domain` couple as one of the primary identifiers, and the only one for which uniqueness is enforced; Mastodon does handle this identifier changing for remote accounts whose protocol identifier stays the same, but the code path for this is inefficient and not widely tested

**Comment by CodingCaro** [comment]
> This needs to be a feature today and the fact that it's not leads me to believe mastodon is not a serious project

Yes.

**Comment by nehemiagurl** [comment]
I recognise that this is hard to implement, but so were post edits and a solution was found. sure, this may be harder than edits, but a request being hard is not an argument against said request and not even an argument for deprioritising it, just an argument for getting more people to think about this.

**Comment by alexgleason** [comment]
Can y'all please fix this so I mention people on nostr like `@alex` instead of `@20cca2632c2649e8aab4626c5225170e9a9dc850109cff81cd8b9abb071d4a36`? This is holding back bridges to other protocols at this point.

**Comment by zopsicle** [comment]
@alexgleason how about you fix it yourself?

**Comment by Ashdemai** [comment]
> [@alexgleason](https://github.com/alexgleason) how about you fix it yourself?

If they would merge his request, I'm sure he would have started working on it already.

**Comment by billstclair** [comment]
Maybe it's best left on the client side, where it can be specialized as pleased. No Nostr-wide agreement necessary.

**Comment by createHernandez** [comment]
Also looking to change a few account usernames. Commenting to follow the progress of this particular issue.

**Comment by ClearlyClaire** [comment]
You don't need to comment to subscribe to an issue. In the future, you can use the “Subscribe” button in the sidebar:
![Image](https://github.com/user-attachments/assets/80bd2114-39e4-4cb1-9dba-8ec0595c4151)

**Comment by ravimakhijapi** [comment]
I think public UUIDs or similar should be used to provide uniqueness to an account instead. The username change history can also be persisted perhaps to address potential misuse.

User name changes can be propagated to servers as needed, I think.

Verifications should then be tied to the UUID and be unique, e.g. a domain verification should be tied to only one UUID, I think.

**Comment by LHMA2025** [comment]
It would be basically impossible since your username acts like for example, an email address. You cannot change it, or you would lose link to all those posts and it would be taken as a new account completely.

**Comment by Ashdemai** [comment]
Not a good example. You can change an email address. See: Gmail https://support.google.com/accounts/answer/19870

**Comment by LHMA2025** [comment]
Ok but like you get the point

**Comment by omentic** [comment]
@LHMA2025, this is not actually the case. ActivityPub provides many capabilities for migrating ownership of objects (like posts and accounts). You may be interested in reading through the following Fediverse Enhancement Proposal that outlines one such approach (intended for cross-instance migration, but also envelopes instance-local username renaming): https://codeberg.org/fediverse/fep/pulls/692

**Comment by flawrist** [comment]
![](https://github.com/user-attachments/assets/1379169d-78da-4b6a-a381-f5acd39f61c5)

Words on the migration screen talk too much, spinning a cheap yarn to calm the nerves. The real trouble waits at the bottom of the form — a glowing button promising only to _move followers_, when a soul wants to _rename an account_ without losing a bit (registration date, content, followers).

X plays that game better. It lets you do exactly what you want. The rest goes without saying.

![](https://github.com/user-attachments/assets/10d5bfbb-d66a-4627-b759-cd179819a776)
